### PR TITLE
feat: Call Popup and Exotel Integration

### PR DESCRIPTION
**Pull request by surajshetty3416**
- **Call Popup** - To show customer/lead information based on the incoming call phone number

call_popup_4

- **Exotel Settings** - To track incoming call from Exotel
Screenshot 2019-06-17 at 11 12 17 AM

- **Communication** (Module) 
    - **Call Log** (DocType) - To store call details
    - **Communication Medium** (DocType) - Medium used for inbound/outbound communication (eg. An exotel phone number (ExoPhone), Skype number, etc.)
        - **Communication Medium Timeslot** (Child DocType)  - To schedule an employee group which will attend the communication medium on a particular timeslot.
Screenshot 2019-06-17 at 10 56 58 AM

### CODE REVIEW

**Comment by surajshetty3416** [comment]
[WIP] docs https://github.com/frappe/erpnext_com/pull/392

File: erpnext/crm/doctype/utils.py [hunk 1]
```diff
@@ -0,0 +1,95 @@
+import frappe
+from frappe import _
+import json
+
+@frappe.whitelist()
+def get_document_with_phone_number(number):
+	# finds contacts and leads
+	if not number: return
+	number = number[-10:]
```
Comment (netchampfaris) [inline erpnext/crm/doctype/utils.py]: Why only 10 digit phone numbers?

File: erpnext/crm/doctype/utils.py [hunk 2]
```diff
@@ -0,0 +1,95 @@
...
+	weekday = frappe.utils.get_weekday()
+
+	available_employee_groups = frappe.db.sql("""SELECT `parent`, `employee_group`
+		FROM `tabCommunication Medium Timeslot`
+		WHERE `day_of_week` = %s
+		AND `parent` = %s
+		AND %s BETWEEN `from_time` AND `to_time`
+	""", (weekday, communication_medium, now_time), as_dict=1)
+	for group in available_employee_groups:
+		employee_emails += [e.user_id for e in frappe.get_doc('Employee Group', group.employee_group).employee_list]
```
Comment (netchampfaris) [inline erpnext/crm/doctype/utils.py]: why not query the child table directly?

File: erpnext/erpnext_integrations/exotel_integration.py [hunk 3]
```diff
@@ -0,0 +1,105 @@
...
+	})
+
+	return response.json()
+
+def get_exotel_settings():
+	return frappe.get_single('Exotel Settings')
+
+@frappe.whitelist(allow_guest=True)
+def get_phone_numbers():
+	numbers = 'some number'
```
Comment (netchampfaris) [inline erpnext/erpnext_integrations/exotel_integration.py]: ?

**Review by netchampfaris — CHANGES_REQUESTED** [review]

File: erpnext/crm/doctype/utils.py [hunk 4]
```diff
@@ -0,0 +1,98 @@
...
+		call_log.summary = summary
+	else:
+		call_log.summary += '<br>' + summary
+	call_log.save(ignore_permissions=True)
+
+def get_employee_emails_for_popup(communication_medium):
+	now_time = frappe.utils.nowtime()
+	weekday = frappe.utils.get_weekday()
+
+	available_employee_groups = frappe.db.sql_list("""SELECT `employee_group`
```
Comment (netchampfaris) [inline erpnext/crm/doctype/utils.py]: `frappe.get_all` ?

File: erpnext/erpnext_integrations/exotel_integration.py [hunk 5]
```diff
@@ -0,0 +1,99 @@
+import frappe
+import requests
+
+# api/method/erpnext.erpnext_integrations.exotel_integration.handle_incoming_call
+# api/method/erpnext.erpnext_integrations.exotel_integration.handle_end_call
+# api/method/erpnext.erpnext_integrations.exotel_integration.handle_missed_call
+
+@frappe.whitelist(allow_guest=True)
+def handle_incoming_call(*args, **kwargs):
```
Comment (netchampfaris) [inline erpnext/erpnext_integrations/exotel_integration.py]: `*args` unused

File: erpnext/erpnext_integrations/exotel_integration.py [hunk 6]
```diff
@@ -0,0 +1,99 @@
...
+		return call_log
+
+def get_call_log(call_payload, create_new_if_not_found=True):
+	call_log = frappe.get_all('Call Log', {
+		'id': call_payload.get('CallSid'),
+	}, limit=1)
+
+	if call_log:
+		return frappe.get_doc('Call Log', call_log[0].name)
+	elif create_new_if_not_found:
```
Comment (netchampfaris) [inline erpnext/erpnext_integrations/exotel_integration.py]: put this part in `create_call_log`

File: erpnext/erpnext_integrations/exotel_integration.py [hunk 7]
```diff
@@ -0,0 +1,99 @@
...
+	call_log = get_call_log(call_payload, False)
+	if call_log:
+		call_log.status = status
+		call_log.duration = call_payload.get('DialCallDuration') or 0
+		call_log.recording_url = call_payload.get('RecordingUrl')
+		call_log.save(ignore_permissions=True)
+		frappe.db.commit()
+		return call_log
+
+def get_call_log(call_payload, create_new_if_not_found=True):
```
Comment (netchampfaris) [inline erpnext/erpnext_integrations/exotel_integration.py]: `get_call_log` should not create a call log

File: erpnext/public/js/call_popup/call_popup.js [hunk 8]
```diff
@@ -0,0 +1,212 @@
...
+		this.dialog.get_close_btn().show();
+		this.dialog.$body.addClass('call-popup');
+		this.dialog.set_secondary_action(this.close_modal.bind(this));
+		frappe.utils.play_sound('incoming-call');
+		this.dialog.show();
+	}
+
+	make_caller_info_section() {
+		const wrapper = this.dialog.fields_dict['caller_info'].$wrapper;
+		wrapper.append('<div class="text-muted"> Loading... </div>');
```
Comment (netchampfaris) [inline erpnext/public/js/call_popup/call_popup.js]: translation

File: erpnext/public/js/call_popup/call_popup.js [hunk 9]
```diff
@@ -0,0 +1,212 @@
...
+	}
+
+	setup_unknown_caller(wrapper) {
+		wrapper.append(`
+			<div class="caller-info">
+				<b>${__('Unknown Number')}:</b> ${this.caller_number}
+				<button
+					class="margin-left btn btn-new btn-default btn-xs"
+					data-doctype="Contact"
+					title="Make New Contact">
```
Comment (netchampfaris) [inline erpnext/public/js/call_popup/call_popup.js]: translation

File: erpnext/public/js/call_popup/call_popup.js [hunk 10]
```diff
@@ -0,0 +1,212 @@
...
+			}
+		}, 10000);
+	}
+
+	make_last_interaction_section() {
+		frappe.xcall('erpnext.crm.doctype.utils.get_last_interaction', {
+			'number': this.caller_number,
+			'reference_doc': this.contact
+		}).then(data => {
+			const comm_field = this.dialog.fields_dict["last_communication"];
```
Comment (netchampfaris) [inline erpnext/public/js/call_popup/call_popup.js]: `this.dialog.get_field('last_communication')`

**Review by netchampfaris — CHANGES_REQUESTED** [review]

**Review by netchampfaris — APPROVED** [review]

**Pull request merged**